### PR TITLE
Add type hints and docstrings to heap.py

### PR DESCRIPTION
**Pull request by MarkHershey**
### **Changes:**

* Add type hints
* Add docstrings
* Add explanatory comments 
* Improve code readability
* Change to use f-string

---

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [x] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by MarkHershey** [comment]
#2510

**Review by 53jk1 — APPROVED** [review]

**Comment by MarkHershey** [comment]
> This pull request has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

This PR still pending approval.

**Comment by dhruvmanila** [comment]
Please run `mypy` on this file on your local machine and check if there are any errors.
```
python -m pip install mypy

mypy --ignore-missing-imports path/to/heap.py
```

**Comment by MarkHershey** [comment]
> Please run `mypy` on this file on your local machine and check if there are any errors.
> 
> ```
> python -m pip install mypy
> 
> mypy --ignore-missing-imports path/to/heap.py
> ```

Done checking, no error.

File: data_structures/heap/heap.py [hunk 1]
```diff
@@ -1,99 +1,131 @@
...
     [201, 107, 25, 103, 100, 15, 1, 9, 7, 5, 11]
     >>>
     >>> h.heap_sort()
-    >>> h.display()
+    >>> print(h)
     [1, 5, 7, 9, 11, 15, 25, 100, 103, 107, 201]
-    >>>
     """
 
     def __init__(self):
```
Comment (dhruvmanila) [inline data_structures/heap/heap.py:27]:
```suggestion
    def __init__(self) -> None:
```

File: data_structures/heap/heap.py [hunk 2]
```diff
@@ -1,99 +1,131 @@
...
-        self.h = []
-        self.curr_size = 0
+        self.h: List[Union[int, float]] = []
```
Comment (dhruvmanila) [inline data_structures/heap/heap.py:28]: `int` and `float` have a subtype relationship and in a union, only the less specific type survives, which means:
```suggestion
        self.h: List[float] = []
```

File: data_structures/heap/heap.py [hunk 3]
```diff
@@ -1,99 +1,131 @@
...
+        self.heap_size: int = 0
+
+    def __repr__(self) -> str:
+        return str(self.h)
+
+    def parent_index(self, child_idx: int) -> Union[int, None]:
```
Comment (dhruvmanila) [inline data_structures/heap/heap.py:34]: As this type is so common, we got a type specifically for this situation:
```suggestion
    def parent_index(self, child_idx: int) -> Optional[int]:
```

File: data_structures/heap/heap.py [hunk 4]
```diff
@@ -1,99 +1,131 @@
...
+        """ return the parent index of given child """
+        if child_idx > 0:
+            return (child_idx - 1) // 2
+        return None
 
-    def get_left_child_index(self, i):
-        left_child_index = 2 * i + 1
-        if left_child_index < self.curr_size:
+    def left_child_idx(self, parent_idx: int) -> Union[int, None]:
```
Comment (dhruvmanila) [inline data_structures/heap/heap.py:40]:
```suggestion
    def left_child_idx(self, parent_idx: int) -> Optional[int]:
```

File: data_structures/heap/heap.py [hunk 5]
```diff
@@ -1,99 +1,131 @@
...
+        """
+        left_child_index = 2 * parent_idx + 1
+        if left_child_index < self.heap_size:
             return left_child_index
         return None
 
-    def get_right_child(self, i):
-        right_child_index = 2 * i + 2
-        if right_child_index < self.curr_size:
+    def right_child_idx(self, parent_idx: int) -> Union[int, None]:
```
Comment (dhruvmanila) [inline data_structures/heap/heap.py:50]:
```suggestion
    def right_child_idx(self, parent_idx: int) -> Optional[int]:
```

File: data_structures/heap/heap.py [hunk 6]
```diff
@@ -1,99 +1,131 @@
...
-                largest = lc
-            if rc is not None and self.h[rc] > self.h[largest]:
-                largest = rc
-            if largest != index:
-                self.h[largest], self.h[index] = self.h[index], self.h[largest]
-                self.max_heapify(largest)
-
-    def build_heap(self, collection):
-        self.curr_size = len(collection)
+    def max_heapify(self, index: int):
```
Comment (dhruvmanila) [inline data_structures/heap/heap.py:60]: This function does not have a return statement which means it implicitly returns `None`.
```suggestion
    def max_heapify(self, index: int) -> None:
```

File: data_structures/heap/heap.py [hunk 7]
```diff
@@ -1,99 +1,131 @@
...
+            if right_child is not None and self.h[right_child] > self.h[violation]:
+                violation = right_child
+            # if violation indeed exists
+            if violation != index:
+                # swap to fix the violation
+                self.h[violation], self.h[index] = self.h[index], self.h[violation]
+                # fix the subsequent violation recursively if any
+                self.max_heapify(violation)
+
+    def build_max_heap(self, collection: Iterable[Union[int, float]]):
```
Comment (dhruvmanila) [inline data_structures/heap/heap.py:80]:
```suggestion
    def build_max_heap(self, collection: Iterable[float]):
```

File: data_structures/heap/heap.py [hunk 8]
```diff
@@ -1,99 +1,131 @@
...
-
-    def get_max(self):
-        if self.curr_size >= 2:
+        self.heap_size = len(self.h)
+        if self.heap_size > 1:
+            # max_heapify from right to left but exclude leaves (last level)
+            for i in range(self.heap_size // 2 - 1, -1, -1):
+                self.max_heapify(i)
+
+    def max(self) -> Union[int, float]:
```
Comment (dhruvmanila) [inline data_structures/heap/heap.py:89]:
```suggestion
    def max(self) -> float:
```

File: data_structures/heap/heap.py [hunk 9]
```diff
@@ -1,99 +1,131 @@
...
+        """ return the max in the heap """
+        if self.heap_size >= 1:
+            return self.h[0]
+        else:
+            raise Exception("Empty heap")
+
+    def extract_max(self) -> Union[int, float]:
```
Comment (dhruvmanila) [inline data_structures/heap/heap.py:96]:
```suggestion
    def extract_max(self) -> float:
```

File: data_structures/heap/heap.py [hunk 10]
```diff
@@ -1,99 +1,131 @@
...
-        elif self.curr_size == 1:
-            self.curr_size -= 1
+        elif self.heap_size == 1:
+            self.heap_size -= 1
             return self.h.pop(-1)
-        return None
+        else:
+            raise Exception("Empty heap")
+
+    def insert(self, value: Union[int, float]):
```
Comment (dhruvmanila) [inline data_structures/heap/heap.py:110]:
```suggestion
    def insert(self, value: float) -> None:
```

File: data_structures/heap/heap.py [hunk 11]
```diff
@@ -1,99 +1,131 @@
...
+        """ insert a new value into the max heap """
+        self.h.append(value)
+        idx = (self.heap_size - 1) // 2
+        self.heap_size += 1
+        while idx >= 0:
+            self.max_heapify(idx)
+            idx = (idx - 1) // 2
 
     def heap_sort(self):
```
Comment (dhruvmanila) [inline data_structures/heap/heap.py:119]:
```suggestion
    def heap_sort(self) -> None:
```

File: data_structures/heap/heap.py [hunk 12]
```diff
@@ -1,99 +1,131 @@
...
-            self.max_heapify(curr)
-            curr = (curr - 1) // 2
+        self.heap_size = size
 
-    def display(self):
-        print(self.h)
 
-
-def main():
+def demo():
```
Comment (dhruvmanila) [inline data_structures/heap/heap.py:128]: Please move the function content in the `__name__ == "__main__":` block, not as a function but as statemens only.

**Review by dhruvmanila — CHANGES_REQUESTED** [review]
This is looking good! 👍 
Please import the type: `from typing import Optional`

File: data_structures/heap/heap.py (hunk 2 above)
Comment (MarkHershey) [reply data_structures/heap/heap.py]: got it, nice to know!

**Review by dhruvmanila — APPROVED** [review]
Awesome work! Thanks for your persistence! LGTM! 🎉

**Pull request merged**